src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs

```diff
@@ -1277,19 +1340,78 @@ private void AddTokenClaims(List<Claim> instanceClaims, TokenInformationClass to
-        TokenImpersonation
+            return SafeImpersonate(_safeTokenHandle);
+        }
+
+        public static WindowsImpersonationContext Impersonate(IntPtr userToken) 
+        {
+            // impersonating a zero token means clear the token on the thread
+            if (userToken == IntPtr.Zero)
+            {
+                return SafeImpersonate(s_invalidTokenHandle);
+            }
+
+            WindowsIdentity wi = new WindowsIdentity(userToken, null, -1);
+            return wi.Impersonate();
+        }
+
+        /// <summary>
+        /// Impersonation workflow:
+        /// 1) We get current usertoken.
+        /// 2) We create WindowsImpersonationContext class with current usertoken and a flag to specify if we'are already in a impersonated context.
+        /// 3) If we'are in a impersonated context WindowsImpersonationContext duplicates current usertoken to allow to revert to correct identity once 
+        ///    impersonation context ends.
+        /// 4) We revert current impersonated context if existent and call ImpersonateLoggedOnUser with new token identity.
+        /// 5) On WindowsImpersonationContext.Undo()/.Dispose() if we were in a impersonated context we revert to correct identity or simply clear the token
+        ///    on the thread.
+        /// </summary>
+        /// <param name="userToken"></param>
+        /// <returns></returns>
+        private static WindowsImpersonationContext SafeImpersonate(SafeAccessTokenHandle userTokenToImpersonate)
+        {
+            SafeAccessTokenHandle currentUserToken = GetCurrentToken(TokenAccessLevels.MaximumAllowed, false, out bool isImpersonating, out int hr);
+
+            if (currentUserToken == null || currentUserToken.IsInvalid)
+            {
+                throw new SecurityException(new Win32Exception().Message);
+            }
+
+            WindowsImpersonationContext context = new WindowsImpersonationContext(currentUserToken, isImpersonating);
+
+            // impersonating a zero token means clear the token on the thread
+            if (userTokenToImpersonate.IsInvalid)
+            { 
+                if (!Interop.Advapi32.RevertToSelf())
+                { 
+                    Environment.FailFast(new Win32Exception().Message);
+                }
+            }
+            // revert existent impersonating token if existent and impersonate with new token identity
+            else
+            {
+                if (!Interop.Advapi32.RevertToSelf())
+                { 
+                    Environment.FailFast(new Win32Exception().Message);
+                }
+
+                if (!Interop.Advapi32.ImpersonateLoggedOnUser(userTokenToImpersonate))
+                {
+                    // if impersonation fails we cleanup WindowsImpersonationContext and throw SecurityException
+                    context.Undo();
+                    throw new SecurityException(SR.Argument_ImpersonateUser);
+                }
+            }
```

Comment:
We're missing the code here that updates ExecutionContext.  That's important.  Is that coming in a subsequent change?  Developers expect to be able to write code like:
```C#
async Task DoWorkAsync(...)
{
    using (var impersonation = windowsIdentity.Impersonate())
    {
        ... // impersonating here
        await ...;
        ... // still impersonating here
    }
}
```
and have the impersonation flow across the await... the ExecutionContext is what makes that possible.

Reply:
>  Is that coming in a subsequent change? 

No I'm aware of it and asked how I can capture and flow without "delegate"(like `RunImpersonate`).
Can you follow my comments above https://github.com/dotnet/corefx/pull/33645#issuecomment-442385027 and enlighten me? Thank's @stephentoub!

Reply:
Why can't we just store the relevant information into an async local, and use the change callback to revert/restore the impersonation, as we do for run impersonated?
https://source.dot.net/#System.Security.Principal.Windows/System/Security/Principal/WindowsIdentity.cs,713

Reply:
Ok I'll work on it, as said above  there'll be some difference between `RunImpersonate/Impersonate` for netcoreapp and netfx...because netfx `RunImpersonate` is a wrapper on `WindowsImpersonationContext` that doesn't capture/revert context(an I don't understand how async/await works on netfx without call to [`ImpersonateLoggedOnUser`](https://referencesource.microsoft.com/#mscorlib/system/security/principal/windowsidentity.cs,756) is we come back on other thread), on current implementation of `RunImpersonate` EC is captured/reverted thank's to `Run`. But there isn't api to revert "current context" that I could capture and use on Undo(), what do you think,  am I wrong?

Reply:
> But there isn't api to revert "current context" that I could capture and use on Undo(), what do you think, am I wrong?

Can you just grab the current value when you (before you) impersonate and then store that value back in Undo?

Reply:
Ok for impersonation token value is not a problem, but for `CultureInfo/SynchronizationContext`....on `RunImpersonate` if I change `CurrentCulture` will be reverted on exit due to `Capture/Run`, to be consistent I should Capture/Restore(this is missing from EC api)